0.2-Elliptic-Curve-Math.ipynb

```diff
@@ -0,0 +1,247 @@
+    "import util\n",
+    "\n",
+    "# Bitcoin Core Test Framework \n",
+    "from test_framework.key import SECP256K1_ORDER, SECP256K1_FIELD_SIZE\n",
+    "from test_framework.key import ECKey, ECPubKey\n",
+    "\n",
+    "# Python Imports\n",
+    "import random\n",
+    "import pdb"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# 0.2 Elliptic Curve Math (Review)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Elliptic Curve math involves scalars and Points(X,Y). \n",
+    "\n",
+    "The following provides an overview of all operations of scalars and points over elliptic curves. In Bitcoin, key pair generation and signing is performed over the SECP256k1 curve.\n",
+    "\n",
+    "![test](images/ec_math0.jpg)\n",
+    "\n",
+    "### Classes / Methods for Elliptic Curve Math\n",
+    "\n",
+    "`int`: All Scalar operations over SECP251k1 can be performed with python integers and the modulo `%` operator:\n",
+    "* All scalar operations are performed modulo n (`SECP256k1_ORDER`).\n",
+    "* Addition: `a + b % SECP256K1_ORDER`\n",
+    "* Subtraction: `-a = SECP256K1_ORDER - a`\n",
+    "* Multiplication: `a * b % % SECP256K1_ORDER`\n",
+    "* Division (Fermat's little Theorem): `1/b = b**(SECP256K1_ORDER-2) % SECP256K1_ORDER`\n",
+    "\n",
+    "`ECKey`: The Bitcoin Core library provides a private key class which can also perform certain scalar operations.\n",
+    "* Addition: `add`\n",
+    "* Subtraction: `add` (adding complement of scalar)\n",
+    "* Multiplication: `mul`\n",
+    "* Division: `mul` (multiplying with scalar inverse, see Fermat's little theorem) \n",
+    "\n",
+    "![test](images/ec_math1.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.1 Example: Scalar Addition over SECP256K1 order\n",
+    "\n",
+    "Addition can be performed with integer modulo math in python or with the private key class `ECKey`. We can set a `ECKey` object to a certain value, or generate a new private key with the `generate` method.\n",
+    "\n",
+    "In the example below, addition is performed with both integers and the `ECKey` class, and evaluated for equality."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# int() operations\n",
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
+    "ab = (a + b) % SECP256K1_ORDER\n",
+    "\n",
+    "# ECkey(operation)\n",
+    "a_key = ECKey()\n",
+    "a_key.set(a.to_bytes(32,'big'), True) \n",
+    "ab_key = a_key.add(b.to_bytes(32,'big'))\n",
+    "\n",
+    "# Ensure equivalency.\n",
+    "print(int().from_bytes(ab_key.get_bytes(),'big') == ab)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.2 Example: Scalar Multiplication over SECP256K1 order\n",
```

Comment:
My additions, matching addition example above but switching for mul:

```
# int() operations
# get 2 random numbers a,b
a = random.randrange(1, SECP256K1_ORDER)
print("a: ", a)
b = random.randrange(1, SECP256K1_ORDER)
print("b: ", b)
# use simple multiplication for a*b but modulo the result to make sure we stay within the SECP256K1_ORDER order
ab = (a * b) % SECP256K1_ORDER
print("a*b (using int operations):   ", ab)

# ECkey() operations
# convert our ints to 32 byte representations since that is what ECKey expects
a_bytes = a.to_bytes(32,'big')
b_bytes = b.to_bytes(32,'big')

# use our bytes to instantiate ECKey instances 
a_key = ECKey() 
a_key.set(a_bytes, True) # set the secret value of the Eckey
b_key = ECKey() 
b_key.set(b_bytes, True) # True means we are using compressed keys

ab_key = a_key.mul(b_key.get_bytes()) # need to pass 32 byte key representation and not an Eckey instance
ab_ec = int().from_bytes(ab_key.get_bytes(),'big') # convert result to int for comparison
print("a*b (using ECKey operations): ", ab_ec)

# Ensure equivalency.
print(ab == ab_ec)
```

Reply:
added. Thanks!